content/6_Advanced/Segtree_Beats.mdx

```diff
@@ -1,21 +1,531 @@
+
+<Info title="Pro Tip">
+
+It's possible to solve this problem without a $\texttt{min}$ tag, but using 
+both a $\texttt{min}$ and a $\texttt{max}$ allows range updates, which 
+isn't strictly pertinent to this particular problem but is still 
+a useful thing to know. 
+
+</Info>
+
+# Implementation - The Child and Sequence
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
+
+const int MAXN = 100000 + 1;  // 1-based
+
+int N;
+ll tsum[MAXN * 4];  // Sum tag
+ll tmax[MAXN * 4];  // Max tag
+ll tmin[MAXN * 4];  // Min tag
+ll lazy[MAXN * 4];  // Lazy updates
+ 
+void pushdown(int t, int tl, int tr) {
+	if (lazy[t] != -1) {
+		int tm = (tl + tr) >> 1;
+ 
+		tsum[t << 1] = lazy[t] * (tm - tl + 1);  // sum set update
+		tmax[t << 1] = lazy[t];  // max set update
+		tmin[t << 1] = lazy[t];  // min set update
+		lazy[t << 1] = lazy[t];  // propogate tag
+ 
+		tsum[t << 1 | 1] = lazy[t] * (tr - tm);  // sum set update
+		tmax[t << 1 | 1] = lazy[t];  // max set update
+		tmin[t << 1 | 1] = lazy[t];  // min set update
+		lazy[t << 1 | 1] = lazy[t];  // propogate tag
+ 
+		lazy[t] = -1;  // reset lazy tag
+	}
+}
+ 
+void update(int l, int r, ll v, int t=1, int tl=1, int tr=N) {
+	if (r < tl || tr < l || tmax[t] < v) {
+		return;
+	}
+	if (l <= tl && tr <= r && tmax[t] == tmin[t]) {
+		ll val = tmax[t] % v;
+		tsum[t] = (tr - tl + 1) * val;
+		tmax[t] = val;
+		tmin[t] = val;
+		lazy[t] = val;
+		return;
+	}
+	pushdown(t, tl, tr);
+ 
+	int tm = (tl + tr) >> 1;
+	update(l, r, v, t << 1, tl, tm);
+	update(l, r, v, t << 1 | 1, tm + 1, tr);
+ 
+	tsum[t] = tsum[t << 1] + tsum[t << 1 | 1];
+	tmax[t] = max(tmax[t << 1], tmax[t << 1 | 1]);
+	tmin[t] = min(tmin[t << 1], tmin[t << 1 | 1]);
+}
+ 
+void update(int i, ll v, int t=1, int tl=1, int tr=N) {
+	if (tl == tr) {
+		tsum[t] = tmax[t] = tmin[t] = v;
+		return;
+	}
+	pushdown(t, tl, tr);
+ 
+	int tm = (tl + tr) >> 1;
+	if (i <= tm) {
+		update(i, v, t << 1, tl, tm);
+	} else {
+		update(i, v, t << 1 | 1, tm + 1, tr);
+	}
+	
+	tsum[t] = tsum[t << 1] + tsum[t << 1 | 1];
+	tmax[t] = max(tmax[t << 1], tmax[t << 1 | 1]);
+	tmin[t] = min(tmin[t << 1], tmin[t << 1 | 1]);
+}
+ 
+ll query(int l, int r, int t=1, int tl=1, int tr=N) {
+	if (r < tl || tr < l) {
+		return 0LL;
+	}
+	if (l <= tl && tr <= r) {
+		return tsum[t];
+	}
+	pushdown(t, tl, tr);
+ 
+	int tm = (tl + tr) >> 1;
+	return query(l, r, t << 1, tl, tm) + query(l, r, t << 1 | 1, tm + 1, tr);
+}
+ 
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	memset(lazy, -1, sizeof(lazy));
+	
+	int Q;
+ 
+	cin >> N >> Q;
+	for (int i = 1; i <= N; i++) {
+		ll a; 
+		cin >> a;
+		update(i, a);
+	}
+	while (Q--) {
+		int t; cin >> t;
+		if (t == 1) {
+			int l, r;
+			cin >> l >> r;
+			cout << query(l, r) << '\n';
+		} else if (t == 2) {
+			int l, r; ll x;
+			cin >> l >> r >> x;
+			update(l, r, x);
+		} else if (t == 3) {
+			int k; ll x;
+			cin >> k >> x;
+			update(k, x);
+		}
+	}
+}
+```
 
 <Resources>
```

Comment:
Start a new section here since this isn't a continuation of the implementation?

Reply:
done